lib/phase/merge.js

```diff
@@ -70,74 +71,79 @@ function handlePipelineTemplateMergeForJobs(parsedDoc, newPipeline, pipelineTemp
+                    field => !ALLOWED_JOB_FIELDS_WITH_PIPELINE_TEMPLATE.includes(field)
+                );
+
+                if (hasExtraField) {
+                    throw new Error(
+                        `Job "${jobName}" has unsupported fields in user yaml. Can only set ${ALLOWED_JOB_FIELDS_WITH_PIPELINE_TEMPLATE}.`
+                    );
                 }
-            });
-
-            // Merge job environment
-            if (oldJob.environment || newJob.environment) {
-                newJob.environment = {
-                    ...newJob.environment,
-                    ...oldJob.environment
-                };
-            }
 
-            // Replace settings
-            const yamlSettingsConfig = oldJob.settings || {};
-            const templateSettingsConfig = newJob.settings || {};
-            const newSlackSettings = yamlSettingsConfig.slack || {};
-            const oldSlackSettings = templateSettingsConfig.slack || {};
-            const newEmailSettings = yamlSettingsConfig.email || {};
-            const oldEmailSettings = templateSettingsConfig.email || {};
-
-            newJob.settings = {
-                ...templateSettingsConfig,
-                ...yamlSettingsConfig
-            };
-            if (!isEmpty(oldSlackSettings)) {
-                newJob.settings = {
-                    ...newJob.settings,
-                    slack: {
-                        ...oldSlackSettings,
-                        ...newSlackSettings
+                // Replace fields
+                fieldsToFlatten.forEach(key => {
+                    if (oldJob[key]) {
+                        newJob[key] = oldJob[key];
                     }
-                };
-            }
-            if (!isEmpty(oldEmailSettings)) {
+                });
+
+                // Merge job environment
+                if (oldJob.environment || newJob.environment) {
+                    newJob.environment = {
+                        ...newJob.environment,
+                        ...oldJob.environment
+                    };
+                }
+
+                // Merge settings
+                const yamlSettingsConfig = oldJob.settings || {};
+                const templateSettingsConfig = newJob.settings || {};
+                const newSlackSettings = yamlSettingsConfig.slack || {};
+                const oldSlackSettings = templateSettingsConfig.slack || {};
+                const newEmailSettings = yamlSettingsConfig.email || {};
+                const oldEmailSettings = templateSettingsConfig.email || {};
+
                 newJob.settings = {
-                    ...newJob.settings,
-                    email: {
-                        ...oldEmailSettings,
-                        ...newEmailSettings
-                    }
+                    ...templateSettingsConfig,
+                    ...yamlSettingsConfig
                 };
-            }
+                if (!isEmpty(oldSlackSettings)) {
```

Comment:
seems redundant, considering we are already setting the `oldSlackSettings` to `{}` based on `templateSettingsConfig.slack` being empty.